### ISSUE TITLE
OpenStack: Floating IPs allocated but not attached to VM

### ISSUE THREAD

**Issue by riccardomurri** [gc3-uzh-ch/elasticluster#497]
As reported by @valyo in https://github.com/gc3-uzh-ch/elasticluster/issues/371#issuecomment-335868407 :

> the frontend node (the one specified by ssh_to=frontend) is not
> getting a floating IP. I don't see anything about assigning floating
> IP in the log, and looking at the openstack dashboard also doesn't
> show any floating IP assigned to frontend or any other node.  I went
> through the documentation but couldn't find anything more than the
> `request_floating_ip` conf key, which I have set to `True`.

**Comment by riccardomurri** [comment]
Indeed, it looks like the code at [lines 477--488 of `providers/openstack.py`][1], contrary to what the comment states, allocates a floating IP but does *not* attach it to the VM.

[1]: https://github.com/gc3-uzh-ch/elasticluster/blob/master/elasticluster/providers/openstack.py#L477

**Comment by valyo** [comment]
I don't know if it's supposed to show up at all, but I don't see a floating IP allocated in the OS dashboard during the cluster startup
Thought it might be useful to mention it.

**Comment by riccardomurri** [comment]
>  I don't see a floating IP allocated in the OS dashboard during the cluster startup

Indeed this does not match my own experiments: I see the IP address is allocated, but not associated.

Can you please post the output of `pip freeze` in your ElastiCluster virtual env?

**Comment by valyo** [comment]
here is the output of `pip freeze`:

```
adal==0.4.7
ansible==2.4.0.0
apache-libcloud==2.2.1
appdirs==1.4.3
asn1crypto==0.23.0
azure==2.0.0
azure-batch==3.0.0
azure-common==1.1.8
azure-datalake-store==0.0.17
azure-graphrbac==0.30.0
azure-keyvault==0.3.7
azure-mgmt==1.0.0
azure-mgmt-authorization==0.30.0
azure-mgmt-batch==4.0.0
azure-mgmt-cdn==0.30.3
azure-mgmt-cognitiveservices==1.0.0
azure-mgmt-compute==1.0.0
azure-mgmt-containerregistry==0.2.1
azure-mgmt-datalake-analytics==0.1.6
azure-mgmt-datalake-nspkg==2.0.0
azure-mgmt-datalake-store==0.1.6
azure-mgmt-devtestlabs==2.0.0
azure-mgmt-dns==1.0.1
azure-mgmt-documentdb==0.1.3
azure-mgmt-iothub==0.2.2
azure-mgmt-keyvault==0.31.0
azure-mgmt-logic==2.1.0
azure-mgmt-monitor==0.2.1
azure-mgmt-network==1.0.0
azure-mgmt-nspkg==2.0.0
azure-mgmt-rdbms==0.1.0
azure-mgmt-redis==4.1.0
azure-mgmt-resource==1.1.0
azure-mgmt-scheduler==1.1.3
azure-mgmt-sql==0.5.3
azure-mgmt-storage==1.0.0
azure-mgmt-trafficmanager==0.30.0
azure-mgmt-web==0.32.0
azure-nspkg==2.0.0
azure-servicebus==0.21.1
azure-servicefabric==5.6.130
azure-servicemanagement-legacy==0.20.6
azure-storage==0.34.3
Babel==2.5.1
bcrypt==3.1.3
boto==2.48.0
certifi==2017.7.27.1
cffi==1.11.1
click==6.7
cliff==2.9.0
cmd2==0.7.7
coloredlogs==7.3
cryptography==2.0.3
debtcollector==1.18.0
-e git://github.com/gc3-uzh-ch/elasticluster.git@7709945d80d05b82274f4593b0895a9e744a1512#egg=elasticluster
enum34==1.1.6
funcsigs==1.0.2
functools32==3.2.3.post2
futures==3.1.1
google-api-python-client==1.6.4
google-compute-engine==2.6.1
httplib2==0.10.3
humanfriendly==4.4.1
idna==2.6
ipaddress==1.0.18
iso8601==0.1.12
isodate==0.5.4
Jinja2==2.9.6
jsonpatch==1.16
jsonpointer==1.12
jsonschema==2.6.0
keyring==10.4.0
keystoneauth1==3.2.0
MarkupSafe==1.0
monotonic==1.3
msgpack-python==0.4.8
msrest==0.4.15
msrestazure==0.4.14
netaddr==0.7.19
netifaces==0.10.6
oauth2client==4.1.2
oauthlib==2.0.4
os-client-config==1.28.0
osc-lib==1.7.0
oslo.config==4.13.1
oslo.i18n==3.18.0
oslo.serialization==2.21.1
oslo.utils==3.30.0
paramiko==2.3.1
pathlib2==2.3.0
pbr==3.1.1
positional==1.2.1
prettytable==0.7.2
pyasn1==0.3.7
pyasn1-modules==0.1.4
pyCLI==2.0.3
pycparser==2.18
PyJWT==1.5.3
PyNaCl==1.1.2
pyOpenSSL==17.3.0
pyparsing==2.2.0
pyperclip==1.5.27
python-cinderclient==3.2.0
python-dateutil==2.6.1
python-gflags==3.1.1
python-glanceclient==2.8.0
python-keystoneclient==3.13.0
python-neutronclient==6.5.0
python-novaclient==9.1.1
pytz==2017.2
PyYAML==3.12
requests==2.14.2
requests-oauthlib==0.8.0
requestsexceptions==1.3.0
rfc3986==1.1.0
rsa==3.4.2
scandir==1.6
schema==0.6.6
simplejson==3.11.1
six==1.11.0
stevedore==1.27.1
unicodecsv==0.14.1
uritemplate==3.0.0
warlock==1.2.0
wrapt==1.10.11
```

**Comment by jehane** [comment]
There is "quick and very dirty" workaround : 
Start your cluster with elasticluster ( elasticluster start $yourcluster )
once the instance are created (when elasticluster fail to connect to instance through private IP) , use ctrl+C to interrupt elasticluster
Manually assign public IP to the instance (through web interface or openstack command line ) 
Restart your cluster (elasticluster start $yourcluster)

**Comment by valyo** [comment]
I did try this and first it seemed like it worked (after assigning IPs manually and restarting elasticluster could ssh to the nodes) but later on through the playbooks I got errors. When I went into the VMs and started to dig in, I found certain issues. It's hard to tell what's the problem but it seems messed up. I hope the floating IP issue will get solvet.
Thanks for the hint anyways, @jehane

**Comment by riccardomurri** [comment]
Hello, just a short note to say that this has not fallen off my radar -- although it has been much harder to investigate and fix than I had imagined.

My current understanding of the issue is this:

* Floating IPs were managed using OpenStack's "novaclient" until recently (version 6.0?) -- after that they were removed and one should use the "neutronclient" instead.  So, a workaround if you really need this functionality quickly is to just downgrade novaclient to an earlier version (e.g., `pip install python-novaclient==6.0.0`)
* Unfortunately, the neutron API alone is not enough for assigning a floating IP to a VM instance.  The complete procedure, as far as I can understand, is this:
  1. Create a floating IP through the "neutron" API.
  2. Find the "port ID" corresponding to the network interface that one wants to assign the floating IP to -- I am still unsure how this should be done, it seems that the "nova" API is needed (port IDs should be returned when one requests info about a VM's network interfaces)
  3. Update the floating IP object gotten in 1., setting the port ID -- this should now associate the floating IP with the given network interface.

I have pushed (unfinished) code to https://github.com/gc3-uzh-ch/elasticluster/pull/507 -- if there is any suggestion on what should go in stead of the placeholder `FIXMEEE` labels, I'll be glad to hear :-)

**Comment by riccardomurri** [comment]
Update: the code in #507 is now finished according to the plan outlined above.  Still, it doesn't work: Neutron refuses to update the floating IP (step *iii.*).  See message accompanying commit https://github.com/gc3-uzh-ch/elasticluster/pull/507/commits/f2fcda821c5a475d54e7b0abefb76854c00d64ad for more info.

Any suggestion is welcome.

**Comment by valyo** [comment]
Hi and thanks for your effort.

I did checkout your branch issues/#496 and tried, and I am getting this error:
`NotFound: Floating IP pool not found.`

With Terraform one needs to specify a Floating IP pool, which is something like "Public External IPv4 Network" in my case, i.e. it's the network label and not the network ID.

BTW, have you considered using Terraform for starting the cluster instances?

**Comment by riccardomurri** [comment]
> I did checkout your branch issues/#496 and tried, and I am getting this error:
>         NotFound: Floating IP pool not found.

Ok, so it looks like an additional configuration parameter is needed,
specify the floating IP pool...

> BTW, have you considered using Terraform for starting the cluster instances?

Not really. Does it have an API callable from Python?

Actually, Ansible has modules for doing all sort of operations with
OpenStack (including starting servers and allocating IP addresses or
ports) -- I gather Terraform would be another external command to do
that.  The problem I have with that is that error reporting is poor:
basically, we can just detect if the external command failed and hope
it gives a clear enough message for the user to fix the problem.
Ansible, very often does not.  Therefore ATM I prefer to do this stuff
in Python where I can control the logging and print info that can help
me debug issues...

But I'm definitely open to alternatives!

**Comment by valyo** [comment]
After a quick look it seems Terraform itself doesn't have a Python API, but I found this one https://github.com/beelit94/python-terraform

We are using Terraform here https://github.com/kubenow/KubeNow and it seems to be very robust and easy to use.
Otherwise, I undertand your concerns about capturing errors from external tools.

Cheers

**Comment by valyo** [comment]
I have another concern regarding floating IPs in Elasticluster, but I'm not sure if I should submit a new issue because it is more like a feature request
ATM Elasticluster needs a floating IP for each node in the cluster, and the floating IPs are not exactly an abundant resource (at least from our current OS cloud provider). 
I was thinking about using Ansible from the master node for configuring the rest of the nodes, but the error reporting problem will be even worse in this case.
What are your thoughts on this?

**Comment by riccardomurri** [comment]
@valyo @jehane Do you know if the OpenStack cloud on which you are trying to allocate floating IPs uses "nova" or "neutron" networking?

**Comment by riccardomurri** [comment]
@valyo  Regarding this comment of yours:

> ATM Elasticluster needs a floating IP for each node in the cluster,
> and the floating IPs are not exactly an abundant resource (at least
> from our current OS cloud provider).  I was thinking about using
> Ansible from the master node for configuring the rest of the nodes,
> but the error reporting problem will be even worse in this case.

It would be possible to tell Ansible to use a "jump host", i.e., to
channel all traffic through a specific host which could be the
front-end.  My experience with this is not great, though: compute
nodes randomly marked as "unreachable", almost impossible to set up a
mid-size cluster without errors (i.e., need to re-run "setup" a few
times over).  I guess that the front-end node cannot keep up with
Ansible setup + forwarding connections...

**Comment by valyo** [comment]
The one I am working with uses Neutron

Cheers
Valentin


On 1 Feb 2018, at 04:28, Riccardo Murri <notifications@github.com<mailto:notifications@github.com>> wrote:


@valyo<https://github.com/valyo> @jehane<https://github.com/jehane> Do you know if the OpenStack cloud on which you are trying to allocate floating IPs uses "nova" or "neutron" networking?

**Comment by riccardomurri** [comment]
This has now been fixed in 9e5c8d8b and 5789b44a, at long last!!

Thank you everyone for your patience.
